swarm/network/simulations/overlay_test.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2018 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p/simulations"
+)
+
+var (
+	nodeCount = 16
+)
+
+//This test is used to test the overlay simulation.
+//As the simulation is executed via a main, it is easily missed on changes
+//An automated test will prevent that
+//The test just connects to the simulations, starts the network,
+//starts the mocker, gets the number of nodes, and stops it again.
+//It also provides a documentation on the steps needed by frontends
+//to use the simulations
+func TestOverlaySim(t *testing.T) {
+	//start the simulation
+	log.Info("Start simulation backend")
+	go runOverlaySim()
+
+	//the simulation runs usually at localhost:8888
+	localhostAddr := fmt.Sprintf("http://localhost:%d", httpSimPort)
+	//wait for the network to start
+	time.Sleep(1 * time.Second)
+
+	log.Debug("Start simulation network")
+	//start the simulation network (initialization of simulation)
+	resp, err := http.Post(localhostAddr+"/start", "application/json", nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		t.Fatalf("Expected Status Code %d, got %d", http.StatusOK, resp.StatusCode)
+	}
+
+	log.Debug("Start mocker")
+	//start the mocker, needs a node count and an ID
+	formstr := []byte(fmt.Sprintf("node-count=%d&mocker-type=%s", nodeCount, simulations.GetMockerList()[0]))
+
+	resp, err = http.Post(localhostAddr+"/mocker/start", "application/x-www-form-urlencoded", bytes.NewReader(formstr))
```

Comment:
`defer resp.Body.Close()` ?

Reply:
yes, but not `defer`, as it's reused later

Reply:
This is a test, so it's not very important. But since this pops up quite often, let me beg to differ, for example:

```
        resp, err = http.Post(localhostAddr+"/mocker/start", "application/x-www-form-urlencoded", bytes.NewReader(formstr))
        if err != nil {
                t.Fatal(err)
        }
        if resp.StatusCode != http.StatusOK {
                reason, err := ioutil.ReadAll(resp.Body)
                if err != nil {
                        t.Fatal(err)
                }
                log.Error(string(reason))
                resp.Body.Close()
                t.Fatalf("Expected Status Code %d, got %d", http.StatusOK, resp.StatusCode)
        }
```

If this fails after the `io.ReadAll`, the body `io.Reader` will stay open.

Actually, we must generally take care when we are reusing the same variable for objects which we perform deferred actions upon. See example 4 on this page:

https://blog.learngoprogramming.com/5-gotchas-of-defer-in-go-golang-part-iii-36a1ab3d6ef1

Reply:
> it's reused later

Note that when you use `defer`, it doesn't matter if you then re-assign the variable later, the deferred call will be on the value that the variable had at the time it was registered with `defer`, for example:

```go
package main

import (
	"fmt"
)

type Closer struct {
	ID string
}

func (c *Closer) Close() {
	fmt.Println("closing:", c.ID)
}

func main() {
	c := &Closer{ID: "1"}
	defer c.Close()

	c = &Closer{ID: "2"}
	defer c.Close()

	c = &Closer{ID: "3"}
	defer c.Close()
}
```

Will print:

```
closing: 3
closing: 2
closing: 1
```

Reply:
ah the above clarifies my issues with it - thanks!